### PR TITLE
Fix for connectors based on refactoring of schemas

### PR DESCRIPTION
**Pull request by ayush-shah**
### Describe your changes :

This fixes Mysql Source connector

#
### Type of change :

- [x] Bug fix
- [x] Improvement
- [x] New feature
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] Documentation

#
### Frontend Preview (Screenshots) :
<p align="center">For frontend related change, please link screenshots of your changes preview! Optional for backend related changes.
</p>

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#
### Reviewers

@harshach @pmbrull

### CODE REVIEW

**Comment by ayush-shah** [comment]
Few things to note:
Sample Data
Table Profiles
Table Queries
Data Model
are not part of CreateTableRequest - Commented those out for now - will add a fix (include them OMetaDatabaseTable model or other approach - create a ticket for the same)
cc: @harshach @pmbrull

File: ingestion/src/metadata/ingestion/sink/metadata_rest.py [hunk 1]
```diff
@@ -101,6 +101,7 @@ def __init__(
     @classmethod
     def create(cls, config_dict: dict, metadata_config: OpenMetadataServerConfig):
         config = MetadataRestSinkConfig.parse_obj(config_dict)
+        metadata_config = OpenMetadataServerConfig.parse_obj(metadata_config)
```
Comment (pmbrull) [inline ingestion/src/metadata/ingestion/sink/metadata_rest.py:104]: This is not needed, we are already passing the class `OpenMetadataServerConfig`, not the dict. As we can read everything typed from the JSON Schemas, we can go and reduce the parsings we needed in the beginning

File: ingestion/src/metadata/ingestion/sink/metadata_rest.py [hunk 2]
```diff
@@ -194,32 +195,32 @@ def write_tables(self, db_schema_and_table: OMetaDatabaseAndTable):
...
-                        sample_data=db_schema_and_table.table.sampleData,
+                        sample_data=db_schema_and_table.table.dict().get("sampleData"),
                     )
                 except Exception as e:
                     logging.error(
                         f"Failed to ingest sample data for table {db_schema_and_table.table.name}"
                     )
 
-            if db_schema_and_table.table.tableProfile is not None:
+            if db_schema_and_table.table.dict().get("tableProfile") is not None:
```
Comment (pmbrull) [inline ingestion/src/metadata/ingestion/sink/metadata_rest.py:209]: Why do we need to `dict` this?

File: ingestion/src/metadata/ingestion/source/mysql.py [hunk 3]
```diff
@@ -19,44 +19,45 @@
...
 class MysqlSource(SQLSource):
     def __init__(self, config, metadata_config):
         super().__init__(config, metadata_config)
 
     @classmethod
-    def create(cls, config_dict, metadata_config: OpenMetadataServerConfig):
-        config = MySQLConfig.parse_obj(config_dict)
+    def create(cls, config_dict, metadata_config_dict):
+        config = MetadataIngestionSourceConfig.parse_obj(config_dict)
+        metadata_config = OpenMetadataServerConfig.parse_obj(metadata_config_dict)
```
Comment (pmbrull) [inline ingestion/src/metadata/ingestion/source/mysql.py:37]: Same as above. Let's keep the `metadata_config: OpenMetadataServerConfig` and use as-is.

**Review by harshach — APPROVED** [review]

File: ingestion/src/metadata/ingestion/source/mysql.py [hunk 4]
```diff
@@ -19,44 +19,45 @@
...
+        # Added parsing of MysqlConnection in order to validate config with pydantic models
+        # Missing / Extra fields will throw errors
+        MysqlConnection.parse_obj(config_dict.get("serviceConnection").get("config"))
+        SourceConfig.parse_obj(config_dict.get("sourceConfig"))
```
Comment (pmbrull) [inline ingestion/src/metadata/ingestion/source/mysql.py:41]: We should not need to recast `SourceConfig` here. Any validation error will be picked up at `config = MetadataIngestionSourceConfig.parse_obj(config_dict)`. For the connection, let's just run a `isinstance` and call it a day. With the schemas we can have great docs to set this up. If the overall structure matches, showing an error if the class is incorrect should be enough

File: ingestion/src/metadata/utils/helpers.py [hunk 5]
```diff
@@ -44,6 +49,39 @@
 logger = logging.getLogger(__name__)
 
 
+@singledispatch
+def get_connection_url(serviceConnectionConfig):
+    raise ValueError(f"Hola {serviceConnectionConfig}")
+
+
+@get_connection_url.register(MysqlConnection)
+def _(serviceConnectionConfig):
```
Comment (pmbrull) [inline ingestion/src/metadata/utils/helpers.py:58]: let's clean this up a bit. We could do something like:

```
@get_connection_url.register
def _(connectionConfig: MysqlConnection):
    ...
```
and just pick the attributes without casting to dict, not needed with proper typing of the arguments

**Review by pmbrull — CHANGES_REQUESTED** [review]

**Comment by pmbrull** [comment]
Let me take a stab on this and we can discuss it together tomorrow 🙏

File: ingestion/src/metadata/ingestion/models/ometa_table_db.py [hunk 6]
```diff
@@ -12,14 +12,14 @@
...
 from metadata.generated.schema.entity.data.databaseSchema import DatabaseSchema
 from metadata.generated.schema.entity.data.location import Location
-from metadata.generated.schema.entity.data.table import Table
 
 
 class OMetaDatabaseAndTable(BaseModel):
     database_schema: DatabaseSchema
     database: Database
-    table: Table
+    table: CreateTableRequest
```
Comment (pmbrull) [inline ingestion/src/metadata/ingestion/models/ometa_table_db.py:24]: We have a ticket open to fix this at another point. Let's leave it for now

Comment (pmbrull) [reply ingestion/src/metadata/ingestion/models/ometa_table_db.py:24]: Once we do that, we'll change the table, db and schema to the `Create` model. But let's not have them mixed

File: ingestion/src/metadata/utils/engines.py [hunk 7]
```diff
@@ -21,6 +21,8 @@
 
 from metadata.generated.schema.metadataIngestion.workflow import Source as SourceConfig
 
+from .helpers import get_connection_url
```
Comment (pmbrull) [inline ingestion/src/metadata/utils/engines.py:24]: also lets try to use full imports

**Comment by pmbrull** [comment]
Implemented in https://github.com/open-metadata/OpenMetadata/pull/3870

**Pull request closed without merging**